channeldb/channel.go

```diff
@@ -1589,6 +1605,83 @@ func (c *OpenChannel) isBorked(chanBucket kvdb.RBucket) (bool, error) {
 	return channel.chanStatus != ChanStatusDefault, nil
 }
 
+// MarkShutdownSent updates the channel status to indicate that a shutdown has
+// been initiated. It also persists the delivery script that we will use in
+// the Shutdown message so that we can guarantee that the same delivery script
+// is used if a re-connect happens.
+func (c *OpenChannel) MarkShutdownSent(localDeliveryScript []byte,
+	locallyInitiated bool) error {
+
+	return c.markShutdownSent(localDeliveryScript, locallyInitiated)
+}
+
+// markShutdownSent is a helper function which modifies the channel status of
+// the receiving channel and inserts a delivery script under the delivery script
+// key. It adds a status which indicates the party that initiated the channel
+// close.
+func (c *OpenChannel) markShutdownSent(deliveryScript []byte,
+	locallyInitiated bool) error {
+
+	c.Lock()
+	defer c.Unlock()
+
+	putDeliveryScript := func(chanBucket kvdb.RwBucket) error {
+		return chanBucket.Put(deliveryScriptKey, deliveryScript)
+	}
+
+	status := ChanStatusShutdownSent
+	if locallyInitiated {
+		status |= ChanStatusLocalCloseInitiator
```

Comment:
We can't add another channel status here because on restart, execution will hit this if statement and the link won't be loaded: https://github.com/lightningnetwork/lnd/blob/cf4f468ab5c70e644d8974145d1ffd74a3dabc35/peer/brontide.go#L881-L882

The reason the itest works is that coop close then proceeds via `restartCoopClose`, but when the coop close transaction is being made it will just delete the HTLC and burn it to fees. This can be seen if you log the coop close transaction and `proposedFee` here: https://github.com/lightningnetwork/lnd/blob/cf4f468ab5c70e644d8974145d1ffd74a3dabc35/lnwallet/channel.go#L7821-L7825

It can also be observed because no `UpdateFulfillHTLC` is sent during the itest.

Instead, we should just be able to use the existence of a delivery script to determine whether or not we need to continue coop close

Reply:
suggestion: I think we should adjust the checks that you're referencing @Crypt-iQ as opposed to refraining from updating statuses according to the events that transpire.

Reply:
The issue is that it isn't default though. We're definitely in a different state that isn't normal operation and can't accept new HTLC adds. Can you clarify what "the risk" is? I'm trying to weigh the cost of having increasingly fragmented logic (where consequences of changing stuff leaks like it is doing now) vs the cost of fixing things so the code straightforwardly reflects what is supposed to happen.

Reply:
There's no issue with using `ChanStatusDefault` here, the link just needs to be aware that it should send `shutdown`. The risk is that we make a costly mistake when changing how the status field is used

Reply:
I think there's a way to navigate this. For one, it's not really functioning as a "status" as it's a bit-vector. Instead it's a collection of flags. So really what the ChanStatusDefault check is doing is trying to check if specific flags are unset, and act accordingly. I think perhaps inverting the check and rather than checking "is it default", we should check "is it not have any one of these conditions". We can always push it to an extra field but I do think that the way we handle ChannelStatus right now is very fragile and should be reworked to be more structurally correct. Maybe an issue for a less tactical PR though.